crates/bevy_ecs/src/schedule/commands.rs

```diff
@@ -0,0 +1,121 @@
+use std::marker::PhantomData;
+
+use super::{IntoSystemDescriptor, Schedule, StageLabel};
+
+#[derive(Default)]
+pub struct ScheduleCommandQueue {
+    items: Vec<Box<dyn ScheduleCommand>>
+}
+
+impl ScheduleCommandQueue {
+    pub fn push<C>(&mut self, command: C)
+    where 
+        C: ScheduleCommand,
+    {
+        self.items.push(Box::new(command));
+    }
+
+    pub fn apply(&mut self, schedule: &mut Schedule) {
+        for command in self.items.drain(..) {
+            command.write(schedule);
+        }
+    }
+
+    pub fn transfer(&mut self, queue: &mut ScheduleCommandQueue) {
+        queue.items.extend(self.items.drain(..));
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+}
+
+/// A [`Schedule`] mutation.
+pub trait ScheduleCommand: Send + Sync + 'static {
+    fn write(self: Box<Self>, schedule: &mut Schedule);
+}
+
+pub struct ScheduleCommands<'a> {
+    queue: &'a mut ScheduleCommandQueue
+}
+
+impl<'a> ScheduleCommands<'a> {
+    pub fn new(queue: &'a mut ScheduleCommandQueue) -> Self {
+        Self {
+            queue
+        }
+    }
+
+    pub fn insert_system<T: 'static, S, Params: 'static>(&mut self, system: T, stage_label: S)
```

Comment:
We should have a simple working example and doc strings here: this is by far the most publicly visible part of the API.